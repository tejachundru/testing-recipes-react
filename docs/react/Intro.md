Working with jest

    npm install --save-dev jest
    npm install --save-dev @testing-library/jest-dom
    npm install --save-dev @testing-library/react

Working with react testing library

    npm install --save-dev @testing-library/react
    npm install --save-dev @testing-library/user-event

Once the dependencies are installed, you can start writing tests for your React components.

## Jest

Jest is a JavaScript testing framework that was developed by Facebook. It is widely used for testing React components and other JavaScript code.

Jest provides a simple and intuitive API for writing tests. It allows you to write tests for your React components in a way that is similar to how you write tests for other JavaScript code.

### Important Syntax and Concepts

- Test Suites: A test suite is a collection of tests that are grouped together. It is used to organize your tests and make them easier to manage.

- Test Cases: A test case is a single test that verifies a specific behavior of your code. It is used to test a specific functionality or feature of your code.

- Test Expectations: A test expectation is an assertion that is used to verify the behavior of your code. It is used to check if a certain condition is true or false.

- Test Runner: A test runner is a tool that is used to run your tests. It is responsible for executing the tests and reporting the results.

### Writing Tests

To write tests for your React components, you can use the `test` , `it` and `describe` functions provided by Jest.

describe - Used to group related tests together.
it - Used to define a test case.
test - Used to define a test case.

Here's an example of a test case with addition of two numbers:

```javascript
const add = (a, b) => a + b;

// describe - Used to group related tests together.
describe("Addition", () => {
  // it - Used to define a test case. Specify the test case name and the expected behavior of the code.
  it("adds two positive numbers", () => {
    const result = add(2, 3);
    expect(result).toBe(5);
  });

  it("adds two negative numbers", () => {
    const result = add(-2, -3);
    expect(result).toBe(-5);
  });

  it("adds two zero numbers", () => {
    const result = add(0, 0);
    expect(result).toBe(0);
  });
});
```

In this example, the `test` function is used to define a test case. The `add` function is used to add two numbers. The `expect` function is used to specify the expected behavior of the code. In this case, the expected behavior is that the result of adding two numbers should be 5.

### Running Tests with Jest

To run the tests, you can use the following command:

```
npm test
```

This command will run the tests and display the results in the console.

### Debugging Tests

If you encounter any issues with your tests, you can use the `jest --watch` command to run the tests in watch mode. This will allow you to see the results of the tests as you make changes to your code.

Here's an example of how to run the tests in watch mode:

```
jest --watch
```

This command will run the tests in watch mode and display the results in the console.

## React Testing Library

React Testing Library is a library that provides a simple and complete API for testing React components. It encourages good testing practices and makes it easier to write tests for React components.

### Important Concepts

- Rendering: Rendering is the process of creating a visual representation of a component in the DOM.
- Testing Library: Testing Library is a library that provides a set of utilities for testing React components.
- User Events: User events are actions that a user can perform on a component, such as clicking a button or typing in a text field.

Recipes With React Testing Library

- [Recipes](react/recipes.md)
