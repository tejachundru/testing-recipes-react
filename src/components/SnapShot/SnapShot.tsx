import React from "react";

const SnapShot = ({ name }: { name?: string }) => {
  return (
    <div>
      <span>Hello {name ? name : "World"}!</span>
    </div>
  );
};

export default SnapShot;
