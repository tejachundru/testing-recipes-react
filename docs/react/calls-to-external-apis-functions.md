## Calls to external APIs/functions

Testing the behavior of a component that calls external APIs/functions is important to ensure that the component behaves as expected and that the expected output is rendered or updated.

To test the behavior of a component that calls external APIs/functions, you can use the `fetch` API or a mocking library like `nock` to simulate the behavior of the external APIs/functions.

Here's an example of testing a component that calls an external API using the `fetch` API:

```jsx
import React from "react";
import { render, screen } from "@testing-library/react";
import { CallsToExternalAPIsFunctions } from "./index";

describe("CallsToExternalAPIsFunctions", () => {
  test("renders the component", async () => {
    render(<CallsToExternalAPIsFunctions />);
    const heading = screen.getByText("Hello, World!");
    const button = screen.getByRole("button");
    expect(heading).toBeInTheDocument();
    expect(button).toBeInTheDocument();

    // Simulate a click on the button
    await screen.findByRole("button").click();

    // Check if the heading text has been updated
    expect(heading).toHaveTextContent("Hello, John!");
  });
});
```

In this example, we are testing a component that calls an external API using the `fetch` API. We simulate a click on the button and then check if the heading text has been updated.

## Reference

- [Testing Library](https://testing-library.com/docs/react-testing-library/api#fetch)
- [React Testing Library](https://testing-library.com/docs/react-testing-library/api#fetch)
