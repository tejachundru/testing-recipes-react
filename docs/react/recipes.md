### 1. Snapshot testing

- Snapshot testing is a way to test the behavior of a component without having to render it.
- It allows you to compare the output of a component with a previously saved snapshot.
- This can be useful when you want to ensure that your component renders the same way as it did in the past.
- To create a snapshot, you can use the `toMatchSnapshot` method provided by Jest.
- Here's an example of how to use snapshot testing in a React component:

```javascript
import React from "react";
import { render } from "@testing-library/react";

const MyComponent = () => {
  return <div>Hello, world!</div>;
};

describe("MyComponent", () => {
  it("renders correctly", () => {
    const { container } = render(<MyComponent />);
    expect(container).toMatchSnapshot();
  });
});
```

In this example, the `render` function is used to render the `MyComponent` component. The `container` property is then used to access the rendered component. The `toMatchSnapshot` method is used to compare the output of the component with a previously saved snapshot.
