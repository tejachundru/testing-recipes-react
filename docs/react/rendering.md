## Rendering UI

Getter functions

- screen.getByText
- screen.getByLabelText
- screen.getByPlaceholderText
- screen.getByRole
- screen.getByTestId
- screen.getByAltText
- screen.getByTitle
  and more...

**_which query should we use?_**
your test should resemble how the user will use/interact with the component as much as possible

---

### Example

```jsx
import React from "react";
import { render, screen } from "@testing-library/react";
import { RenderingComponents } from "./index";

describe("RenderingComponents", () => {
  test("renders the component", () => {
    render(
      <RenderingComponents
        title="Hello, this is a title"
        description="This is a description"
      />
    );
    const heading = screen.getByText("Hello, World!");
    const input = screen.getByPlaceholderText("Type something");
    const button = screen.getByRole("button");
    expect(heading).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });
});
```
