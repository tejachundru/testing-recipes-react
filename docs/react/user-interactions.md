## User Interactions

Testing user interactions is a crucial part of testing. It ensures that the component behaves as expected when the user interacts with it.

There are different ways to test user interactions, such as:

- Simulating user events
- Testing the component's behavior

**_simulating user events_**

Simulating user events involves sending input to the component, such as clicking a button or typing in a text field. This allows you to test how the component responds to different user actions.

**_testing the component's behavior_**

Testing the component's behavior involves checking if the component behaves as expected when the user interacts with it. This can be done by checking if the component renders correctly, if the component responds correctly to user actions, and if the component displays the expected output.

---

### Example

```jsx
import React from "react";
import { render, screen } from "@testing-library/react";
import { HoverEvent } from "./index";

describe("HoverEvent", () => {
  test("renders the component", () => {
    render(<HoverEvent />);
    const tooltip = screen.getByText(/hover over me/i);
    const tooltipText = screen.getByText(/tooltip text/i);
    expect(tooltip).toBeInTheDocument();
    expect(tooltipText).toBeInTheDocument();
  });
});
```

---

### Example

```jsx
import React from "react";
import { render, screen } from "@testing-library/react";
import { HoverEvent } from "./index";

describe("HoverEvent", () => {
  test("renders the component", () => {
    render(<HoverEvent />);
    const tooltip = screen.getByText(/hover over me/i);
    const tooltipText = screen.getByText(/tooltip text/i);
    expect(tooltip).toBeInTheDocument();
    expect(tooltipText).toBeInTheDocument();
  });
});
```

### what did we achieve?

- we have a test that renders the component
- we have a test that checks if the tooltip is visible
  In general, we have a test that checks if the component is rendered correctly
  with expected values
