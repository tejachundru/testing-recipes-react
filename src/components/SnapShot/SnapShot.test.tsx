import { render } from "@testing-library/react";
import React from "react";
import SnapShot from "./SnapShot";

describe("SnapShot", () => {
  it("renders correctly without no props", () => {
    const { container } = render(<SnapShot />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with props", () => {
    const { container } = render(<SnapShot name="John" />);
    expect(container).toMatchSnapshot();
  });
});
