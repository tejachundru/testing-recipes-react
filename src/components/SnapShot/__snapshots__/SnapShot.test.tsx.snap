// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapShot renders correctly with props 1`] = `
<div>
  <div>
    <span>
      Hello 
      John
      !
    </span>
  </div>
</div>
`;

exports[`SnapShot renders correctly without no props 1`] = `
<div>
  <div>
    <span>
      Hello 
      World
      !
    </span>
  </div>
</div>
`;
